Flaky failure in test / Run Unit Tests with Race Detector

```
2023-07-07T15:02:43.254Z	ERROR	header/sync	sync/sync_head.go:150	invalid network header	{"height_of_invalid": 4, "hash_of_invalid": "020839B2619512EB3536A335EF0266E16F5A355BE8790CDF7708D149166BD504", "height_of_subjective": 1, "hash_of_subjective": "00981389EFCA48D4122326EB58292DEB65F1FF6720B7D02ABAB2EAB8218532A4", "reason": "last header hash E985492AA8ECB1EB5BAC21BBC75C01A1040404AE983D544B425D678CA6A7FC3A does not match hash of previous header 00981389EFCA48D4122326EB58292DEB65F1FF6720B7D02ABAB2EAB8218532A4"}
--- FAIL: TestFraudProofService (6.39s)
    full_node_integration_test.go:438: 
        	Error Trace:	/home/runner/work/rollkit/rollkit/node/full_node_integration_test.go:438
        	            				/home/runner/work/rollkit/rollkit/node/full_node_integration_test.go:450
        	Error:      	Received unexpected error:
        	            	datastore: key not found
        	Test:       	TestFraudProofService
```

--- a/mempool/v1/mempool_test.go
+++ b/mempool/v1/mempool_test.go
@@ -236,7 +236,7 @@ func testReapMaxBytesMaxGas(t *testing.T, txmp *TxMempool, tTxs []testTx, expect
 	require.Len(t, reapedTxs, 50)
 
 	// reap by transaction bytes only
-	reapedTxs = txmp.ReapMaxBytesMaxGas(1000, -1)
+	reapedTxs = txmp.ReapMaxBytesMaxGas(1500, -1)
 	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap by transaction bytes only"))
 	require.Equal(t, len(tTxs), txmp.Size())
 	require.Equal(t, expectedSize, txmp.SizeBytes())

--- a/mempool/v1/mempool_test.go
+++ b/mempool/v1/mempool_test.go
@@ -598,7 +598,6 @@ func TestTxMempool_allEntriesSorted(t *testing.T) {
 
 	// Call allEntriesSorted to get the WrappedTx objects sorted by priority and timestamp
 	sorted := txmp.allEntriesSorted()
-	fmt.Println(sorted)
 
 	// Check that the WrappedTx objects are sorted correctly
 	require.Equal(t, wt2, sorted[0])

--- a/mempool/v1/mempool_test.go
+++ b/mempool/v1/mempool_test.go
@@ -15,6 +15,7 @@ import (
 	"github.com/cometbft/cometbft/types"
 
 	"github.com/rollkit/rollkit/mempool"
+	test "github.com/rollkit/rollkit/test/log"
 )
 
 func TestTxMempool_TxsAvailable(t *testing.T) {
@@ -191,11 +192,13 @@ func TestTxMempool_Flush(t *testing.T) {
 	require.Equal(t, int64(0), txmp.SizeBytes())
 }
 
-func TestTxMempool_ReapMaxBytesMaxGas(t *testing.T) {
-	txmp := setup(t, 0)
+func TestTxMempool_Reap(t *testing.T) {
+	// Create mempool and transactions
+	txmp := setupCustom(t, 0, test.NewFileLogger(t))
 	tTxs := checkTxs(t, txmp, 100, 0) // all txs request 1 gas unit
 	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
+	expectedSize := int64(5690)
+	require.Equal(t, expectedSize, txmp.SizeBytes())
 
 	txMap := make(map[types.TxKey]testTx)
 	priorities := make([]int64, len(tTxs))
@@ -209,36 +212,66 @@ func TestTxMempool_ReapMaxBytesMaxGas(t *testing.T) {
 		return priorities[i] > priorities[j]
 	})
 
-	ensurePrioritized := func(reapedTxs types.Txs) {
+	ensurePrioritized := func(reapedTxs types.Txs, testCase string) {
 		reapedPriorities := make([]int64, len(reapedTxs))
 		for i, rTx := range reapedTxs {
 			reapedPriorities[i] = txMap[rTx.Key()].priority
 		}
 
-		require.Equal(t, priorities[:len(reapedPriorities)], reapedPriorities)
+		require.Equal(t, priorities[:len(reapedPriorities)], reapedPriorities, testCase)
 	}
-
+	t.Run("ReapMaxBytesMaxGas", func(t *testing.T) {
+		testReapMaxBytesMaxGas(t, txmp, tTxs, expectedSize, ensurePrioritized)
+	})
+	t.Run("ReapMaxTxs", func(t *testing.T) {
+		testReapMaxTxs(t, txmp, tTxs, expectedSize, ensurePrioritized)
+	})
+}
+func testReapMaxBytesMaxGas(t *testing.T, txmp *TxMempool, tTxs []testTx, expectedSize int64, checkFunc func(reapedTxs types.Txs, testCase string)) {
 	// reap by gas capacity only
 	reapedTxs := txmp.ReapMaxBytesMaxGas(-1, 50)
-	ensurePrioritized(reapedTxs)
+	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap by gas capacity only"))
 	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
+	require.Equal(t, expectedSize, txmp.SizeBytes())
 	require.Len(t, reapedTxs, 50)
 
 	// reap by transaction bytes only
 	reapedTxs = txmp.ReapMaxBytesMaxGas(1000, -1)
-	ensurePrioritized(reapedTxs)
+	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap by transaction bytes only"))
 	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
+	require.Equal(t, expectedSize, txmp.SizeBytes())
 	require.GreaterOrEqual(t, len(reapedTxs), 16)
 
 	// Reap by both transaction bytes and gas, where the size yields 31 reaped
 	// transactions and the gas limit reaps 25 transactions.
 	reapedTxs = txmp.ReapMaxBytesMaxGas(1500, 30)
-	ensurePrioritized(reapedTxs)
+	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap by both transaction bytes and gas"))
 	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
+	require.Equal(t, expectedSize, txmp.SizeBytes())
 	require.Len(t, reapedTxs, 25)
+}
+
+func testReapMaxTxs(t *testing.T, txmp *TxMempool, tTxs []testTx, expectedSize int64, checkFunc func(reapedTxs types.Txs, testCase string)) {
+	// reap all transactions
+	reapedTxs := txmp.ReapMaxTxs(-1)
+	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap all transactions"))
+	require.Equal(t, len(tTxs), txmp.Size())
+	require.Equal(t, expectedSize, txmp.SizeBytes())
+	require.Len(t, reapedTxs, len(tTxs))
+
+	// reap a single transaction
+	reapedTxs = txmp.ReapMaxTxs(1)
+	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap a single transaction"))
+	require.Equal(t, len(tTxs), txmp.Size())
+	require.Equal(t, expectedSize, txmp.SizeBytes())
+	require.Len(t, reapedTxs, 1)
+
+	// reap half of the transactions
+	reapedTxs = txmp.ReapMaxTxs(len(tTxs) / 2)
+	checkFunc(reapedTxs, fmt.Sprint(t.Name(), "reap half of the transactions"))
+	require.Equal(t, len(tTxs), txmp.Size())
+	require.Equal(t, expectedSize, txmp.SizeBytes())
+	require.Len(t, reapedTxs, len(tTxs)/2)
 }
 
 func TestTxMempoolTxLargerThanMaxBytes(t *testing.T) {
@@ -261,55 +294,6 @@ func TestTxMempoolTxLargerThanMaxBytes(t *testing.T) {
 	reapedTxs := txmp.ReapMaxBytesMaxGas(100, -1)
 	require.Len(t, reapedTxs, 1)
 	require.Equal(t, types.Tx(smallTx), reapedTxs[0])
-}
-
-func TestTxMempool_ReapMaxTxs(t *testing.T) {
-	txmp := setup(t, 0)
-	tTxs := checkTxs(t, txmp, 100, 0)
-	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
-
-	txMap := make(map[types.TxKey]testTx)
-	priorities := make([]int64, len(tTxs))
-	for i, tTx := range tTxs {
-		txMap[tTx.tx.Key()] = tTx
-		priorities[i] = tTx.priority
-	}
-
-	sort.Slice(priorities, func(i, j int) bool {
-		// sort by priority, i.e. decreasing order
-		return priorities[i] > priorities[j]
-	})
-
-	ensurePrioritized := func(reapedTxs types.Txs) {
-		reapedPriorities := make([]int64, len(reapedTxs))
-		for i, rTx := range reapedTxs {
-			reapedPriorities[i] = txMap[rTx.Key()].priority
-		}
-
-		require.Equal(t, priorities[:len(reapedPriorities)], reapedPriorities)
-	}
-
-	// reap all transactions
-	reapedTxs := txmp.ReapMaxTxs(-1)
-	ensurePrioritized(reapedTxs)
-	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
-	require.Len(t, reapedTxs, len(tTxs))
-
-	// reap a single transaction
-	reapedTxs = txmp.ReapMaxTxs(1)
-	ensurePrioritized(reapedTxs)
-	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
-	require.Len(t, reapedTxs, 1)
-
-	// reap half of the transactions
-	reapedTxs = txmp.ReapMaxTxs(len(tTxs) / 2)
-	ensurePrioritized(reapedTxs)
-	require.Equal(t, len(tTxs), txmp.Size())
-	require.Equal(t, int64(5690), txmp.SizeBytes())
-	require.Len(t, reapedTxs, len(tTxs)/2)
 }
 
 func TestTxMempool_CheckTxExceedsMaxSize(t *testing.T) {
@@ -562,4 +546,64 @@ func TestTxMempool_CheckTxPostCheckError(t *testing.T) {
 			require.NoError(t, txmp.CheckTx(tx, callback, mempool.TxInfo{SenderID: 0}))
 		})
 	}
+}
+
+func TestTxMempool_allEntriesSorted(t *testing.T) {
+	txmp := setup(t, 0)
+	txs := checkTxs(t, txmp, 5, 0)
+
+	// Create some WrappedTx objects with different priorities and timestamps
+	//
+	// wt2 wt4 and wt5 should be sorted first because it has the highest priority, even though they have a later timestamp than wt1
+	// wt2 wt4 wt5 should keep the same ordering because they have the same priority
+	// wt1 should be sorted fourth because it has the same priority as wt3 but an earlier timestamp
+	// wt3 should be sorted last because it has the same priority as wt1 but a later timestamp
+	wt1 := &WrappedTx{
+		tx:        txs[0].tx,
+		timestamp: time.Now().Add(-time.Hour),
+		priority:  1,
+		sender:    "sender1",
+	}
+	wt2 := &WrappedTx{
+		tx:        txs[1].tx,
+		timestamp: time.Now(),
+		priority:  2,
+		sender:    "sender2",
+	}
+	wt3 := &WrappedTx{
+		tx:        txs[2].tx,
+		timestamp: time.Now(),
+		priority:  1,
+		sender:    "sender3",
+	}
+	wt4 := &WrappedTx{
+		tx:        txs[3].tx,
+		timestamp: time.Now(),
+		priority:  2,
+		sender:    "sender4",
+	}
+	wt5 := &WrappedTx{
+		tx:        txs[4].tx,
+		timestamp: time.Now(),
+		priority:  2,
+		sender:    "sender5",
+	}
+
+	// Add the WrappedTx objects to the TxMempool
+	txmp.insertTx(wt1)
+	txmp.insertTx(wt2)
+	txmp.insertTx(wt3)
+	txmp.insertTx(wt4)
+	txmp.insertTx(wt5)
+
+	// Call allEntriesSorted to get the WrappedTx objects sorted by priority and timestamp
+	sorted := txmp.allEntriesSorted()
+	fmt.Println(sorted)
+
+	// Check that the WrappedTx objects are sorted correctly
+	require.Equal(t, wt2, sorted[0])
+	require.Equal(t, wt4, sorted[1])
+	require.Equal(t, wt5, sorted[2])
+	require.Equal(t, wt1, sorted[3])
+	require.Equal(t, wt3, sorted[4])
 }

--- a/mempool/v1/mempool_test_helpers.go
+++ b/mempool/v1/mempool_test_helpers.go
@@ -69,6 +69,10 @@ func (app *application) CheckTx(req abci.RequestCheckTx) abci.ResponseCheckTx {
 }
 
 func setup(t testing.TB, cacheSize int, options ...TxMempoolOption) *TxMempool {
+	return setupCustom(t, cacheSize, log.TestingLogger().With("test", t.Name()), options...)
+}
+
+func setupCustom(t testing.TB, cacheSize int, log log.Logger, options ...TxMempoolOption) *TxMempool {
 	t.Helper()
 
 	app := &application{kvstore.NewApplication()}
@@ -86,7 +90,7 @@ func setup(t testing.TB, cacheSize int, options ...TxMempoolOption) *TxMempool {
 		require.NoError(t, appConnMem.Stop())
 	})
 
-	return NewTxMempool(log.TestingLogger().With("test", t.Name()), cfg.Mempool, appConnMem, 0, options...)
+	return NewTxMempool(log, cfg.Mempool, appConnMem, 0, options...)
 }
 
 // mustCheckTx invokes txmp.CheckTx for the given transaction and waits until

